### PR TITLE
Add support for language and region to autogenerated schemes

### PR DESCRIPTION
**Pull request by olejnjak**
### Short description 📝

During development I often run into issues that tests fail because of different regions or languages on various simulators (especially local dev machine and CI). This made me think that it would be great if we were capable of setting language and region of autogenerated schemes.

I was thinking more that extending `TestingOptions` struct would be better to adding new assoc values, but as it needs to conform to `RawRepresentable` it would be messy.

### How to test the changes locally 🧐

Just create a project that uses autogenerated schemes options for both workspace and projects.

### Contributor checklist ✅

- [x] The code has been linted using run `./fourier lint tuist --fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Contributors have checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

**Review by kwridan — COMMENTED** [review]
Thanks @olejnjak 

Question, should this perhaps be `language` and `region` and apply to both test and run? rather than only for test? 

Mind extending one of the fixtures to demo the usage of those new options?

**Comment by olejnjak** [comment]
> Thanks @olejnjak
> 
> Question, should this perhaps be `language` and `region` and apply to both test and run? rather than only for test?

It depends I feel more like if it should be added also for run action, it should have separate properties.

> Mind extending one of the fixtures to demo the usage of those new options?

Okay.

**Comment by danyf90** [comment]
> It depends I feel more like if it should be added also for run action, it should have separate properties.

When do you think they would be different?

**Comment by olejnjak** [comment]
> When do you think they would be different?

I pretty often use predefined pseudolanguages for running while when snapshotting UI in tests I prefer "natural" language. Also if you set language only, you get locale without region, which might not be problem for testing, for running app I am not so sure..

**Comment by danyf90** [comment]
> Maybe I would do it this way - introduce a struct, let's call it `SchemeLocale`, in ProjectDescription, with lang and region properties, then to autogenerated schemes options, there would be two properties (`runLocale` and `testLocale`) that would allow users to specify different locale for run and test actions but in case they want to have both with the same locale, they could simply define a constant. This would also simplify sharing locales between workspace and project schemes.
> 
> What do you think guys?

given changing the pair of variables to a `SchemaLocale` struct everywhere could be breaking change, I would just keep the separate variables and add two for the run as well for now

**Comment by danyf90** [comment]
Set to draft while discussing what to do for the run action

**Comment by kwridan** [comment]
> given changing the pair of variables to a SchemaLocale struct everywhere could be breaking change, I would just keep the separate variables and add two for the run as well for now

Perhaps the breaking changes could be mitigated if it's only scoped to the newly added API on `AutomaticSchemesOptions`  as opposed to changing it everywhere?

**Comment by danyf90** [comment]
> > given changing the pair of variables to a SchemaLocale struct everywhere could be breaking change, I would just keep the separate variables and add two for the run as well for now
> 
> Perhaps the breaking changes could be mitigated if it's only scoped to the newly added API on `AutomaticSchemesOptions` as opposed to changing it everywhere?

Given it's as it is (i.e. without struct) everywhere, I would keep it the same here, WDYT?

**Comment by kwridan** [comment]
Makes sense 👍 

Keeps things consistent, we could revisit this in the next major update.

**Comment by danyf90** [comment]
@olejnjak if you add the parameters for run as well we are good to go I guess! 🎉

**Comment by alexanderwe** [comment]
@olejnjak I faced this issue today, just to find out there is an PR which was just opened. Nice to see this already being tackled. Looking forward to it

**Comment by olejnjak** [comment]
Yes, will do during the weekend 🙃

**Comment by olejnjak** [comment]
Should be ready, not sure why that single check is failing 🤔

File: Sources/ProjectDescription/WorkspaceGenerationOptions.swift [hunk 1]
```diff
@@ -18,7 +18,14 @@ extension Workspace {
...
             case disabled
             /// Tuist will generate schemes with the associated testing options
-            case enabled(codeCoverageMode: CodeCoverageMode = .disabled, testingOptions: TestingOptions = [])
+            case enabled(
+                codeCoverageMode: CodeCoverageMode = .disabled,
+                testingOptions: TestingOptions = [],
+                testLanguage: SchemeLanguage? = nil,
+                testRegion: String? = nil,
+                runLanguage: SchemeLanguage? = nil,
+                runRegion: String? = nil
```
Comment (kwridan) [inline Sources/ProjectDescription/WorkspaceGenerationOptions.swift:27]: Looking at the code, we can omit `runLanguage` and `runRegion` from the workspace generation options as the workspace scheme doesn't include/generate a run action 🤐

It took me a while to figure out why setting those options wasn't working 🤦‍♂️

File: Sources/TuistGraph/Models/RunActionOptions.swift [hunk 2]
```diff
@@ -6,6 +6,9 @@ public struct RunActionOptions: Equatable, Codable {
     /// App Language.
     public let language: String?
 
+    /// App Region.
+    public let region: String?
```
Comment (kwridan) [inline Sources/TuistGraph/Models/RunActionOptions.swift:10]: ouch this wasn't wired in all this time - good find 👍 

I think we're missing mapping this property in `RunActionOptions+ManifestMapper`

**Review by kwridan — COMMENTED** [review]
Thanks for the updates @olejnjak

**Review by danyf90 — APPROVED** [review]
Nothing to add to Kas' comment

File: Sources/ProjectDescription/WorkspaceGenerationOptions.swift (hunk 1 above)
Comment (olejnjak) [reply Sources/ProjectDescription/WorkspaceGenerationOptions.swift:27]: Yeah my bad, even though I was thinking about it 🤦‍♂️

**Review by danyf90 — APPROVED** [review]

File: projects/tuist/fixtures/project_with_test_and_run_locales/Project.swift [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+            runRegion: "US"
+        )
+    ),
+    targets: [
+        .init(
+            name: "App",
+            platform: .iOS,
+            product: .app,
+            bundleId: "io.tuist.app"
+        ),
```
Comment (kwridan) [inline projects/tuist/fixtures/project_with_test_and_run_locales/Project.swift:19]: minor: I believe a test target is needed in order the test action to be added

```suggestion
        ),
        .init(
            name: "AppTests",
            platform: .iOS,
            product: .unitTests,
            bundleId: "io.tuist.appTests"
        ),
```

**Review by kwridan — APPROVED** [review]
Awesome thank you @olejnjak 👍

File: Sources/TuistGraph/Models/RunActionOptions.swift (hunk 2 above)
Comment (olejnjak) [reply Sources/TuistGraph/Models/RunActionOptions.swift:10]: Good catch, just please double check I am not sure it is all that should be done in https://github.com/tuist/tuist/pull/4983/commits/c88168ca1e56b1d01ed5e0724e4f76db6a6e452d

**Review by kwridan — APPROVED** [review]
Thanks @olejnjak

**Comment by olejnjak** [comment]
Can running from my fork be the reason why [cloud acceptance test](https://github.com/tuist/tuist/actions/runs/4004301591/jobs/6874055701) job is failing?

**Comment by danyf90** [comment]
> Can running from my fork be the reason why [cloud acceptance test](https://github.com/tuist/tuist/actions/runs/4004301591/jobs/6874055701) job is failing?

Should be! and mighty @fortmarek fixed this, I've aligned with main and it should pass now 🤞

**Comment by JohnnyTseng** [comment]
Thanks @olejnjak for implementing the `testLanguage` feature, I just ran into exact same problem on tuist v3.15.0
Then I found this awesome PR that solved my problem. 

Thank you!

**Comment by JohnnyTseng** [comment]
Thanks @olejnjak for implementing the `testLanguage` feature, I just ran into exact same problem on tuist v3.15.0
Then I found this awesome PR that solved my problem. 

Thank you!

**Pull request merged**